---
output: github_document
---

# Why ?

Because...

Un petit document qui expliquent certains choix de développement faits.



## Pourquoi ne pas utiliser `checkmate::makeAssertionFunction()`

J'aurai aimé juste construire les fonctions `check` et créer les `assert` avec. Seulement, le nom des objets est mal pris en compte pour les `check`s.

Un petit exemple illustratif :

```{r, echo = TRUE, eval = TRUE, error = TRUE}
library("checkmate")
```

```{r, echo = TRUE, eval = TRUE, error = TRUE}
# Un check un peu random fait à la main
check_inf_2 <- function(x, .var.name = checkmate::vname(x)) {
    cond1 <- check_integerish(x, lower = 1, upper = 2)
    return(ifelse(isTRUE(cond1), TRUE, paste("*", .var.name, ":", cond1)))
}

assert_inf_2 <- makeAssertionFunction(check_inf_2)

print(check_inf_2(1, .var.name = "Nom bizarre"))
print(check_inf_2(3, .var.name = "Nom bizarre"))

assert_inf_2(1, .var.name = "Nom bizarre")
assert_inf_2(3, .var.name = "Nom bizarre")
```

On remarque directement qu'on ne peut pas créer de `check` avec `.var.name` puis créer une `assert` à partir de ça.

Un autre possibilité serait de ne pas utiliser de `.var.name` pour les `check`s mais :

```{r, echo = TRUE, eval = TRUE, error = TRUE}
# Un check sans .var.name (donc "normal")
check_inf_2 <- function(x) {
    return(check_integerish(x, lower = 1, upper = 2))
}

check_objet <- function(x) {
    cond1 <- check_inf_2(x)
    cond2 <- check_inf_2(length(x))
    
    output <- c()
    verif <- TRUE
    
    if (!isTRUE(cond1)) {
        output <- c(output, cond1)
        verif <- FALSE
    }
    
    if (!isTRUE(cond2)) {
        output <- c(output, cond2)
        verif <- FALSE
    }
    
    output <- paste("*", output, collapse = "\n")
    
    return(ifelse(verif, verif, output))
}

assert_objet <- makeAssertionFunction(check_objet)
```

```{r, echo = TRUE, eval = TRUE, error = TRUE}
check_objet(4)
check_objet(c(1, 1, 2))
```

Remarque :
La fonction de `check` marche bien mais uniquement parce qu'on l'a doté de `.var.name`.

```{r, echo = TRUE, eval = TRUE, error = TRUE}
assert_objet(4)
assert_objet(c(1, 1, 2))
```

Mais on remarque alors que le message d'erreur n'est pas très clair...

Pire :

```{r, echo = TRUE, eval = TRUE, error = TRUE}
assert_objet(c(4, 1, 2))
```

Dernière solution : faire des fonctions mixtes :

```{r, echo = TRUE, eval = TRUE, error = TRUE}
# Un check avec .var.name
check_inf_2 <- function(x, .var.name = checkmate::vname(x)) {
    cond1 <- check_integerish(x, lower = 1, upper = 2)
    return(ifelse(isTRUE(cond1), TRUE, paste("*", .var.name, ":", cond1)))
}

# Un check sns
check_objet <- function(x) {
    cond1 <- check_inf_2(x, .var.name = "x")
    cond2 <- check_inf_2(length(x), .var.name = "length(x)")
    
    output <- c()
    verif <- TRUE
    
    if (!isTRUE(cond1)) {
        output <- c(output, cond1)
        verif <- FALSE
    }
    
    if (!isTRUE(cond2)) {
        output <- c(output, cond2)
        verif <- FALSE
    }
    
    output <- paste(output, collapse = "\n")
    
    return(ifelse(verif, verif, output))
}

assert_objet <- makeAssertionFunction(check_objet)
```

```{r, echo = TRUE, eval = TRUE, error = TRUE}
print(check_objet(c(1, 1, 2)))
assert_objet(c(1, 1, 2))

assert_objet(c(4, 1, 2))
```

Mais cette solution n'est pas viable à long terme car on aimerait quand même pouvoir transformer en `assert` nos `check` (pour `check_inf_2`) et peut être que `check_objet` sera aussi appelé par une autre fonction et ne fournira alors à son tour aucune info sur les objets scannés.



## Pourquoi faire une faible utilisation de `checkmate::makeAssertCollection()` ?

La fonction `checkmate::makeAssertCollection()` permet de regrouper des assert ensemble et de mettre un message d'erreur commun plutôt que d'en envoyer un dès qu'une erreur est levée.

Par exemple :

```{r, echo = TRUE, eval = TRUE, error = TRUE}
library("checkmate")

f <- function(x) {
    coll <- checkmate::makeAssertCollection()
    
    checkmate::assert_integer(x, add = coll)
    checkmate::assert_count(x, positive = TRUE, add = coll)
    
    checkmate::reportAssertions(coll)
    
    return(x + 1)
}
```

```{r, echo = TRUE, eval = TRUE, error = TRUE}
f(1L)
f(1.)
f(-1L)
```

Mais lorsque les checks concernent plusieurs variables, il peut y a voir des erreurs dans les erreurs !
... (à compléter au besoin...)
